Allows {{image}} var to be used in policies

Santosh1176 opened the pull request:
Signed-off-by: Santosh Kaluskar <dtshbl@gmail.com>

## Explanation
This PR creates a new variable `{{image}}` to be used in Policies. 
<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue
Closes #4436 

<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## What type of PR is this
Bug Fix
<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests
The following is the result after applying [Verify Images with Multiple Keys](https://main.kyverno.io/policies/other/verify_image_with_multi_keys/) policy.

Rules in the policy applied:
```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: verify-image-with-multi-keys
  annotations:
    policies.kyverno.io/title: Verify Image with Multiple Keys
    policies.kyverno.io/category: Sample
    policies.kyverno.io/severity: medium
    policies.kyverno.io/subject: Pod
    policies.kyverno.io/minversion: 1.7.0
    kyverno.io/kyverno-version: 1.7.2
    kyverno.io/kubernetes-version: "1.23"
    policies.kyverno.io/description: >-
      There may be multiple keys used to sign images based on
      the parties involved in the creation process. This image
      verification policy requires the named image be signed by
      two separate keys. It will search for a global "production"
      key in a ConfigMap called `key` in the `default` Namespace
      and also a Namespace key in the same ConfigMap.
spec:
  validationFailureAction: enforce
  background: true
  rules:
    - name: check-image-with-two-keys
      match:
        any:
        - resources:
            kinds:
              - Pod
      context:
      - name: keys
        configMap:
          name: keys
          namespace: default
      verifyImages:
        # check global key
        - image: "*"
          key: "{{ keys.data.org }}"
        # check image specific key - lookup via image name
        - image: "ghcr.io/stone-nucleus-366711/*"
          key: "{{ keys.data.{{image.name}}}}"
```
The resource, I have applied against the policy:
```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: nginx1
  name: nginx1
  namespace: default
spec:
  containers:
  - image: gcr.io/stone-nucleus-366711/nginx-test
    name: nginx
    envFrom:
      - configMapRef:
          name: keys
    resources: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Always
status: {}
```

ConfigMap I am using:
```yaml
 santosh@shell-scripts*$:cat my-cm.yaml 
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: default
  name: keys
data:
  org: |-
    -----BEGIN PUBLIC KEY-----
    MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEkooBXoWI+9fAJWeWoB26K539sTp/
    50J9t2brN73cxQURl1TCbUvw+3T/XmOCwVrkP6stjHJN2SatnhLmx6736A==
    -----END PUBLIC KEY-----    
  org1:
    -----BEGIN PUBLIC KEY-----
    MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEkooBXoWI+9fAJWeWoB26K539sTp/
    50J9t2brN73cxQURl1TCbUvw+3T/XmOCwVrkP6stjHJN2SatnhLmx6736A==
    -----END PUBLIC KEY-----
```


<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [x] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

shahpratikr commented:
If I could comment, @Santosh1176 could you please try to create a resource with this policy? I tried to look at this issue but I was not able to create resources with only these changes.

Santosh1176 commented:
@shahpratikr Thanks for checking this out, I'll take a look at it. Meanwhile, can you share the error you are getting while creating a resource.

shahpratikr commented:
I was seeing `image.name` not found error. I ran `kubectl run test --image <image>` command to create resource.

Santosh1176 commented:
> I was seeing image.name not found error

```bash
santosh@shell-scripts*$:k apply -f nginx.yaml 
Error from server: error when creating "nginx.yaml": admission webhook "mutate.kyverno.svc-fail" denied the request: 

policy Pod/default/nginx for resource violations: 

check-image:
  check-image: |
    failed to verify image docker.io/nginx:latest: .attestors[0].entries[0].keys: no matching signatures:
    ; .attestors[0].entries[1].keys: no matching signatures:
verify-image-with-multi-keys:
  check-image-with-two-keys: 'failed to substitute variables: failed to resolve keys.data.org
    at path /verifyImages/0/key: JMESPath query failed: Unknown key "org" in path'
```

I am getting an error for `Unkown key` and no- matching signatures, as I've not yet signed the images and not about `image.name`.

JimBugwadia commented:
@Santosh1176 - can you try a test with a signed image?

realshuting commented:
@Santosh1176 - I don't see the variable {{image}} is specified in this policy [Verify Images with Multiple Keys](https://main.kyverno.io/policies/other/verify_image_with_multi_keys/), can you attach full manifests for testing?

Santosh1176 commented:
@realshuting Sorry my bad, That was due to testing with the `check-image` policy, which after testing with the correct image is working as expected. But now I am unable to apply the `verify-image-with-multi-keys` policy ( which was successful as noted in Proof Manifest in my first post). I've changed the imageReference as:
```yaml
verifyImages:
      - imageReferences:
        - "ghcr.io/stone-nucleus-366711/nginx:latest"
        required: true
```
Error while applying Policy:
```bash
k apply -f verify-image-with-multi-keys.yaml 
Error from server: error when creating "verify-image-with-multi-keys.yaml": admission webhook "validate-policy.kyverno.svc" denied the request: spec.rules[0].verifyImages[0].attestors[0]: Invalid value: v1.AttestorSet{Count:(*int)(0xc002035ab8), Entries:[]v1.Attestor{v1.Attestor{Keys:(*v1.StaticKeyAttestor)(0xc00041fe80), Certificates:(*v1.CertificateAttestor)(nil), Keyless:(*v1.KeylessAttestor)(nil), Attestor:(*v1.JSON)(nil), Annotations:map[string]string(nil), Repository:""}, v1.Attestor{Keys:(*v1.StaticKeyAttestor)(0xc00041fec0), Certificates:(*v1.CertificateAttestor)(nil), Keyless:(*v1.KeylessAttestor)(nil), Attestor:(*v1.JSON)(nil), Annotations:map[string]string(nil), Repository:""}}}: Only one entry is currently supported
```



my configmap looks like this:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: default
  name: keys
data:
  allkeys: |-
    -----BEGIN PUBLIC KEY-----
    MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEkooBXoWI+9fAJWeWoB26K539sTp/
    50J9t2brN73cxQURl1TCbUvw+3T/XmOCwVrkP6stjHJN2SatnhLmx6736A==
    -----END PUBLIC KEY-----    
    -----BEGIN PUBLIC KEY-----
    MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEkooBXoWI+9fAJWeWoB26K539sTp/
    50J9t2brN73cxQURl1TCbUvw+3T/XmOCwVrkP6stjHJN2SatnhLmx6736A==
    -----END PUBLIC KEY-----
```
I've tried to test with same key, as I am not sure about signing images with multiple keys.

realshuting commented:
I'm a bit lost here. Per the issue description we need to support variable `{{image}}` in the imageVerify policy, see policy snippet Chip attached [here](https://github.com/kyverno/policies/issues/376#issuecomment-1229562541):
```
      verifyImages:
        # check global key
        - image: "*"
          key: "{{ keys.data.org }}"
        # check image specific key - lookup via image name
        - image: "ghcr.io/chipzoller/*"
          key: "{{ keys.data.{{image.name}}}}"
```

@Santosh1176 - however I failed to find this variable in neither of you testing policies, am I missing anything?

Santosh1176 commented:
@realshuting Ahh.. I was using the policy you mentioned. Ive modified it as in the original, Now I am getting the error that @shahpratikr was mentioning.
```bash
verify-image-with-multi-keys:
  check-image-with-two-keys: 'failed to substitute variables: failed to resolve image.name
    at path /verifyImages/1/key: JMESPath query failed: Unknown key "image" in path'
```

I'll take a closer look at this, thanks.

Santosh1176 commented:
@realshuting After adding the `name` also as a var, the request for resource creation is accepted. But, as I've some signature issues, I am getting `imagePullBackoff` error.
```bash
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  2m42s                default-scheduler  Successfully assigned default/nginx to kind-control-plane
  Normal   Pulling    56s (x4 over 2m41s)  kubelet            Pulling image "gcr.io/stone-nucleus-366711/nginx-test"
  Warning  Failed     52s (x4 over 2m37s)  kubelet            Failed to pull image "gcr.io/stone-nucleus-366711/nginx-test": rpc error: code = Unknown desc = failed to pull and unpack image "gcr.io/stone-nucleus-366711/nginx-test:latest": failed to resolve reference "gcr.io/stone-nucleus-366711/nginx-test:latest": pulling from host gcr.io failed with status code [manifests latest]: 401 Unauthorized
  Warning  Failed     52s (x4 over 2m37s)  kubelet            Error: ErrImagePull
  Warning  Failed     39s (x6 over 2m37s)  kubelet            Error: ImagePullBackOff
  Normal   BackOff    27s (x7 over 2m37s)  kubelet            Back-off pulling image "gcr.io/stone-nucleus-366711/nginx-test"
```

Santosh1176 commented:
I made changes to the vars, and now the request to create the resource is successful:
```bash
 santosh@shell-scripts*$:k describe cpol verify-image-with-multi-keys 
Name:         verify-image-with-multi-keys
Namespace:    
Labels:       <none>
<snip>
.
.
<snpi>
 Name:  check-image-with-two-keys
    Verify Images:
      Image:                  *
      Key:                    {{ keys.data.org }}
      Mutate Digest:          true
      Required:               true
      Verify Digest:          true
      Image:                  ghcr.io/stone-nucleus-366711/*
      Key:                    {{ keys.data.{{image.name}}}}
      Mutate Digest:          true
      Required:               true
      Verify Digest:          true
  Validation Failure Action:  enforce
Events:                       <none>
```
when I create a pod and it is accepted:
```bash
santosh@shell-scripts*$:k get pod
NAME    READY   STATUS             RESTARTS   AGE
nginx   0/1     ImagePullBackOff   0          8m45s
 santosh@shell-scripts*$:k describe pod nginx 
Name:         nginx
Namespace:    default
Priority:     0
Node:         kind-control-plane/172.20.0.3
Start Time:   Thu, 27 Oct 2022 16:59:31 +0530
<snip>
.
.
<snip>
Events:
  Type     Reason     Age                     From               Message
  ----     ------     ----                    ----               -------
  Normal   Scheduled  8m52s                   default-scheduler  Successfully assigned default/nginx to kind-control-plane
  Normal   Pulling    7m6s (x4 over 8m52s)    kubelet            Pulling image "gcr.io/stone-nucleus-366711/nginx-test"
  Warning  Failed     7m2s (x4 over 8m48s)    kubelet            Failed to pull image "gcr.io/stone-nucleus-366711/nginx-test": rpc error: code = Unknown desc = failed to pull and unpack image "gcr.io/stone-nucleus-366711/nginx-test:latest": failed to resolve reference "gcr.io/stone-nucleus-366711/nginx-test:latest": pulling from host gcr.io failed with status code [manifests latest]: 401 Unauthorized
  Warning  Failed     7m2s (x4 over 8m48s)    kubelet            Error: ErrImagePull
  Warning  Failed     6m48s (x6 over 8m47s)   kubelet            Error: ImagePullBackOff
  Normal   BackOff    3m49s (x19 over 8m47s)  kubelet            Back-off pulling image "gcr.io/stone-nucleus-366711/nginx-test"
```
The `imagePullBackOff` error is due to some Auth issues with gcr.
@realshuting please review.
Thank you!

JimBugwadia commented:
> I made changes to the vars, and now the request to create the resource is successful:

Can you please clarify what changes you made?

Also, please post the policy as a YAML instead of the text / describe output.

And finally, please update your repo branch with the base main.

Santosh1176 commented:
@JimBugwadia I've addded `image` to `wildCardAllowedVariables`.

Policy I am using:

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: verify-image-with-multi-keys
  annotations:
    policies.kyverno.io/title: Verify Image with Multiple Keys
    policies.kyverno.io/category: Sample
    policies.kyverno.io/severity: medium
    policies.kyverno.io/subject: Pod
    policies.kyverno.io/minversion: 1.7.0
    kyverno.io/kyverno-version: 1.7.2
    kyverno.io/kubernetes-version: "1.23"
    policies.kyverno.io/description: >-
      There may be multiple keys used to sign images based on
      the parties involved in the creation process. This image
      verification policy requires the named image be signed by
      two separate keys. It will search for a global "production"
      key in a ConfigMap called `key` in the `default` Namespace
      and also a Namespace key in the same ConfigMap.
spec:
  validationFailureAction: enforce
  background: true
  rules:
    - name: check-image-with-two-keys
      match:
        any:
        - resources:
            kinds:
              - Pod
      context:
      - name: keys
        configMap:
          name: keys
          namespace: default
      verifyImages:
        # check global key
        - image: "*"
          key: "{{ keys.data.org }}"
        # check image specific key - lookup via image name
        - image: "ghcr.io/stone-nucleus-366711/*"
          key: "{{ keys.data.{{image.name}}}}"
```
I've updated and rebased.

chipzoller commented:
Please also update your initial post in the Proof Manifests section with not only the rule but a test resource we can use (which uses publicly-accessible, signed images) to verify the operation of this PR.

Santosh1176 commented:
> Please also update your initial post in the Proof Manifests section with not only the rule but a test resource we can use (which uses publicly-accessible, signed images) to verify the operation of this PR.

Done, I hope it is as desired.

realshuting commented:
@Santosh1176 - thanks for the update! Can you also share your configmap `default/keys` definition?

Santosh1176 commented:
@realshuting Sorry for the delay, I am facing some inconsistency in the output while testing the changes in debug mode. Whenever I restart my system and enter the [debug mode](https://github.com/kyverno/kyverno/blob/main/DEVELOPMENT.md#debugging-local-code). The request for creating the resource is accepted and gets created for the first time. Later after deleting the policy and Pod, any attempt to create a resource fails. I might be doing something wrong, I am looking into it.
Meanwhile, I am updating the changes to the one I made in my first commit.

Current commit accepts {{image}} as var and creates policy.

realshuting commented:
@Santosh1176 - if you can provide more details including policy and resource manifests you used, I can help troubleshoot the issue.

Santosh1176 commented:
@realshuting Thanks for helping me out, I've attached all the manifests I am using in the [first post](https://github.com/kyverno/kyverno/pull/5122#issue-1419887805)

realshuting commented:
> @realshuting Thanks for helping me out, I've attached all the manifests I am using in the [first post](https://github.com/kyverno/kyverno/pull/5122#issue-1419887805)

Thanks! Do you have a test image that I can run locally?

Santosh1176 commented:
> > @realshuting Thanks for helping me out, I've attached all the manifests I am using in the [first post](https://github.com/kyverno/kyverno/pull/5122#issue-1419887805)
> 
> Thanks! Do you have a test image that I can run locally?

[Image I am testing with](gcr.io/stone-nucleus-366711/nginx-test)

Santosh1176 commented:
Hey @realshuting, am I missing something important here?

Also can you please point me to some resources on where to look for the added var to be accessible while evaluating the policy?

chipzoller commented:
Moving to 1.9, can possibly move to a 1.8.4 if there are other items.

What's needed to close this out? Any further discussion?

chipzoller commented:
Yes, it's supposed to at least be substituted with the full value of the `image` field for the container object being processed.

realshuting commented:
> Yes, it's supposed to at least be substituted with the full value of the `image` field for the container object being processed.

The issue https://github.com/kyverno/kyverno/issues/4436 is to allow policies with `image` variable to be accepted, but I'm not sure if this variable ever works before. We should either reject the policy or have it work completely (1. accept the variable; 2. resolve the variable correctly).

@Santosh1176 - what's your thoughts?

chipzoller commented:
@JimBugwadia this did work at some point, correct? I assume you would not have authored the policy otherwise.

JimBugwadia commented:
Yes, the `image` variable works but only in the `verifyImages` rule. 

Here is an example:

https://github.com/kyverno/demos/blob/f4aaa337699bfeebc8a5b55b4a20a6a420b4d805/vulnerabilities/check-scan_cm.yaml
